### PR TITLE
Added sklearn as an extra and optional dependancy

### PR DESCRIPTION
**Pull request by marko-polo-cheno**
Adds optional dependancy (mainly for sklearn)

### CODE REVIEW

File: kolena/_utils/scikit-learn.py
Comment (ly-kolena) [inline kolena/_utils/scikit-learn.py]: What is the purpose of this file?

File: README.md
Comment (ly-kolena) [inline README.md:26]: This looks like instruction for development. Should we include the instruction for users wanting to install kolena through `pip`?

File: kolena/_utils/scikit-learn.py
Comment (marko-polo-cheno) [reply kolena/_utils/scikit-learn.py]: To check if sklearn is installed or not in advance, [similar to this](https://github.com/kolenaIO/kolena/blob/trunk/kolena/_utils/datatypes.py#L33). I might have misunderstood what @gordonhart had in mind.

Comment (ly-kolena) [reply kolena/_utils/scikit-learn.py]: I think this may have to be part of your object detection PR, which would use the code here to check sklearn is installed if users want to use (import) anything in `_experimental.object_detection`.

File: README.md
Comment (ly-kolena) [reply README.md:26]: Example 7 [here](https://pip.pypa.io/en/stable/cli/pip_install/#examples)

File: .circleci/continue_config.yml
Comment (ly-kolena) [inline .circleci/continue_config.yml]: I would change this to `ci-base-<< matrix.python-version >>-<< matrix.extras >>` and remove the `-` from `extras` value. (or `ci-base-<< matrix.extras >>-<< matrix.python-version >>`, to me, display name is less of a concern compared with dealing with extra dash in the value)

note: references also need to be updated.

File: README.md
Comment (gordonhart) [inline README.md]: Remove backticks — quotes around this shouldn't be necessary in common shells, and backticks are separate thing entirely.

File: kolena/_extras/metrics/sklearn.py
Comment (gordonhart) [inline kolena/_extras/metrics/sklearn.py]: I would remove this in favor of suppressing the error that this exists to quiet — something like `# noqa: F401` on the import line

File: kolena/_extras/metrics/sklearn.py
Comment (gordonhart) [inline kolena/_extras/metrics/sklearn.py:18]: Are the quotes around `kolena[metrics]` required?

File: tests/unit_extras/__init__.py
Comment (gordonhart) [inline tests/unit_extras/__init__.py:1]: @ly-kolena what do you think of this unit testing setup?

File: .circleci/continue_config.yml
Comment (ly-kolena) [inline .circleci/continue_config.yml]: For consistency, please also update the `requires` and/or `matrix.extras` to integration test and other jobs that also depends on `ci-base`.

**Review by gordonhart — APPROVED** [review]
Looks nice! We'll want to split `kolena/_extras/metrics/__init__.py` apart at some point such that we can lazily import libraries as required, but this is great to get started.

Will want to get approval from @ly-kolena on the testing approach before merging.

File: tests/unit/_extras/test_extras_fail.py
Comment (ly-kolena) [inline tests/unit/_extras/test_extras_fail.py]: Nit: since this is to test import fail check, can remove these parametrizations.

File: kolena/_extras/metrics/sklearn.py
Comment (marko-polo-cheno) [reply kolena/_extras/metrics/sklearn.py:18]: maybe not - I think at some point I ran into an error (didn't appear again) asking me to use quotes or `kolena\[metrics\]`

File: tests/unit_extras/__init__.py
Comment (ly-kolena) [reply tests/unit_extras/__init__.py:1]: What do people think of using marker `metrics` to highlight tests that need matching extras dependency?

If things get tedious, we can use some pytest helper function to mark whole directory/test-file with a marker.

Comment (ly-kolena) [reply tests/unit_extras/__init__.py:1]: Unless it's a concern to restructure unit-tests later, I'm also fine with refining the structure later when we integrate `sklearn` into new and/or existing workflows.

Comment (gordonhart) [reply tests/unit_extras/__init__.py:1]: > What do people think of using marker metrics to highlight tests that need matching extras dependency?

I like this a lot. Easily extensible to additional extras when those become necessary.

Comment (ly-kolena) [reply tests/unit_extras/__init__.py:1]: If going with the marker route, unit-test structure could look like this:

```
├── tests/
│   ├── unit/
│       ├── _extras/
│          ├── __init__.py
│          └── test_metrics.py
```

and we can put tests from `test_extras_fail.py` and `test_extras_metrics.py` together into `test_metrics.py`
```
def test__extras__metrics_import_check() -> None:
       with pytest.raises(ImportError):
        from kolena._extras.metrics import sklearn_metrics

        sklearn_metrics.precision_recall_fscore_support(np.array(["cat", "dog"]), np.array(["cat", "pig"]),
            average="macro")

@pytest.mark.metrics
@pytest.mark.parametrize(
    "y_true, y_pred, precision, recall",
    [
        (
            np.array(["cat", "dog", "pig", "cat", "dog", "pig"]),
            np.array(["cat", "pig", "dog", "cat", "cat", "dog"]),
            2 / 9,
            1 / 3,
        ),
    ],
)
def test__extras__sklearn__pr(y_true: np.ndarray, y_pred: np.ndarray, precision: float, recall: float) -> None:
    prec, rec, _, _ = sklearn_metrics.precision_recall_fscore_support(y_true, y_pred, average="macro")
    assert precision == pytest.approx(prec, abs=1e-5)
    assert recall == pytest.approx(rec, abs=1e-5)
```

CI can be adjusted
```
poetry run pytest -m "not metrics"
poetry run pytest -m "metrics"
```
respectively.

File: .circleci/continue_config.yml
Comment (ly-kolena) [inline .circleci/continue_config.yml]: Nit: do we need both or one of them suffice?

Comment (marko-polo-cheno) [reply .circleci/continue_config.yml]: ah I see, `requires` structures change, but not the actual key

**Review by ly-kolena — APPROVED** [review]

File: .circleci/continue_config.yml
Comment (gordonhart) [reply .circleci/continue_config.yml]: Is this dangling hyphen a result of this change?

Screen Shot 2023-06-07 at 9 47 45 AM

Comment (marko-polo-cheno) [reply .circleci/continue_config.yml]: Apologies for the delay! Yes, because the extra-input is either `""` or `"metrics"` and the key is like `f"ci-base-{version}-{extra}"`

Comment (gordonhart) [reply .circleci/continue_config.yml]: I know that @ly-kolena has pointed in the other direction, but I'd like to avoid this dangling hyphen in these CI jobs. It's a low priority for sure

**Pull request merged**
